# Baseline modeling

**GOAL:**

- Train multiple models
- compare their performance and metrics
- select the promising model

In [2]:
# Imports

import pandas as pd
import numpy as np 


In [3]:
# load X_train, X_val, X_test, y_train, y_val, y_test

X_train = pd.read_csv(r'..\Data\processed\modeling\X_train.csv')
X_val = pd.read_csv(r'..\Data\processed\modeling\X_val.csv')
X_test = pd.read_csv(r'..\Data\processed\modeling\X_test.csv')

y_train = pd.read_csv(r'..\Data\processed\modeling\y_train.csv')
y_val = pd.read_csv(r'..\Data\processed\modeling\y_val.csv')
y_test = pd.read_csv(r'..\Data\processed\modeling\y_test.csv')

## 1: Baseline model

- linear regression

**comparison models**

- ridge regression
- random forest
- XG boost

**Comparison**

- compare by RMSE Rsqt

In [4]:
# 1.1 Baseline model selcetion (Linear regression)
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

lr.fit(X_train, y_train)
lr_y_pred = lr.predict(X_val)

In [15]:
# 1.2 Modle evaluation (Error metric and Eval metric)
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error, root_mean_squared_error

# Regression error metrics
lr_mae = mean_absolute_error(y_val, lr_y_pred)
lr_mse = mean_squared_error(y_val, lr_y_pred)
lr_rmse = root_mean_squared_error(y_val, lr_y_pred)
print('Error metrics for log transformed rent\n')
print(f'MAE of Linear Regression is: {mean_absolute_error(y_val, lr_y_pred)}')
print(f'MSE of Linear Regression is: {lr_mse}')
print(f'RMSE of Linear Regression is: {lr_rmse}')

# regression evaluation metric
lr_r2 = r2_score(y_val, lr_y_pred)

print(f'\nR2 Score of Linear Regression is: {lr_r2}')

# actual rent metrics

actual_y_val = np.expm1(y_val)
actual_lr_y_pred = np.expm1(lr_y_pred)

# Regression error metrics
lr_mae = mean_absolute_error(actual_y_val, actual_lr_y_pred)
lr_mse = mean_squared_error(actual_y_val, actual_lr_y_pred)
lr_rmse = root_mean_squared_error(actual_y_val, actual_lr_y_pred)

print('\nError metrics for actual rent\n')
print(f'MAE of Linear Regression is: {lr_mae}')
print(f'MSE of Linear Regression is: {lr_mse}')
print(f'RMSE of Linear Regression is: {lr_rmse}')

# regression evaluation metric

print(f'\nR2 Score of Linear Regression is: {lr_r2}')

Error metrics for log transformed rent

MAE of Linear Regression is: 0.1994128873538294
MSE of Linear Regression is: 0.09182113494594465
RMSE of Linear Regression is: 0.3030200240016238

R2 Score of Linear Regression is: 0.3344733140396472

Error metrics for actual rent

MAE of Linear Regression is: 1754.580646621875
MSE of Linear Regression is: 19910211.309875313
RMSE of Linear Regression is: 4462.085981900765

R2 Score of Linear Regression is: 0.3344733140396472


# Baseline Model Report

## MAE - Error Metric
- Mean Absolute Error in actual price,
    - tells me that my regression model's predictions are **on average, ~1754 rs** away from the actual rent

## RMSE - Error Metric
- RMSE tells a gap,
    ```md
    RMSE = 4462
    MAE  = 1754
    Gap  = 2708  ← this is large
    ```
    - This gap means my model has some bad predictions (outliers), that pulls up RMSE score
    - Most predictions are around ₹1,754 off (MAE)
    - But a few are severely bad — pulling RMSE up to ₹4,462
    - The gap (₹2,708) just tells us how much those outliers are skewing the score

## R2 Score - Eval metric

- R2 Score tells me that my baseline model (Linear Regression), can explains about ~33% of variation in the rent
    - so Since R2 score was low (below 50% means) the model is weak

- ** why R2 score is low?**
    - Linear regression can not predict non linear predictions

# Final Observation: Linear Regression Baseline

---

| Metric | Score | Indication | Better | Decision |
|--------|-------|------------|--------|----------|
| **MAE** | ₹1,754 | On average, every prediction is ₹1,754 away from the actual rent | ⬇️ Lower = Better | Train another model and compare |
| **RMSE > MAE** | Gap = ₹2,708 | Some predictions are severely bad (outliers) | ⬇️ Lower = Better | Train another model and compare |
| **R² Score** | 0.33 | Linear Regression model's prediction is weak — ~67% of rent variation unexplained | ⬆️ Higher = Better | Train another model (tree-based) to improve |

---

> **Verdict:** Linear Regression is not good enough. Move to tree-based models (Random Forest / XGBoost) to capture non-linear rent patterns.